rust/src/fvm/machine.rs

```diff
@@ -307,3 +333,148 @@ fn import_actors(
     assert_eq!(roots.len(), 1);
     Ok(Some(roots[0]))
 }
+
+#[derive(Clone, Debug, Serialize_tuple, Deserialize_tuple)]
+struct LotusTrace {
+    pub msg: Message,
+    pub msg_receipt: Receipt,
+    pub error: String,
+    pub subcalls: Vec<LotusTrace>,
+}
+
+fn build_lotus_trace(
+    new_call: &ExecutionEvent,
+    trace_iter: &mut impl Iterator<Item = ExecutionEvent>,
+) -> anyhow::Result<LotusTrace> {
+    let mut new_trace = LotusTrace {
+        msg: match new_call {
+            ExecutionEvent::Call(send_params) => Message {
+                version: 0,
+                from: Address::new_id(send_params.from),
+                to: send_params.to,
+                sequence: 0,
+                value: send_params.value.clone(),
+                method_num: send_params.method,
+                params: send_params.params.clone(),
+                gas_limit: 0,
+                gas_fee_cap: TokenAmount::default(),
+                gas_premium: TokenAmount::default(),
+            },
+            _ => {
+                return Err(anyhow!("expected ExecutionEvent of type Call"));
+            }
+        },
+        msg_receipt: Receipt {
+            exit_code: ExitCode::OK,
+            return_data: RawBytes::default(),
+            gas_used: 0,
+        },
+        error: String::new(),
+        subcalls: vec![],
+    };
+
+    while let Some(trace) = trace_iter.next() {
```

Comment:
This pretty much repeats a bulk of the logic in the ref-fvm Executor. This layer should not be forming receipts, but _receiving_ already-formed receipts. Also, replicating the error number to exit code mapping table for top-level calls is a smell.

Reply:
This was intentional to keep the FVM clean. Lotus expects Receipts, but receipts are for top-level messages, not sub-calls. So we convert to a format that lotus expects inside the FFI, with the goal of eventually fixing Lotus.